gslib/tests/test_wrapped_credentials.py

```diff
@@ -0,0 +1,78 @@
+# -*- coding: utf-8 -*-
+# Copyright 2020 Google Inc. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Tests for wrapped_credentials.py."""
+
+from google.auth import credentials
+from gslib.tests import testcase
+from gslib.utils.wrapped_credentials import WrappedCredentials
+import httplib2
+import oauth2client
+from oauth2client.contrib import dictionary_storage
+
+from six import add_move, MovedModule
+
+add_move(MovedModule("mock", "mock", "unittest.mock"))
+from six.moves import mock
+
+ACCESS_TOKEN = "foo"
+CONTENT = "content"
+RESPONSE = httplib2.Response({
+    "content-type": "text/plain",
+    "status": "200",
+    "content-length": len(CONTENT),
+})
+
+
+class MockCredentials(credentials.Credentials):
+  refresh = mock.Mock()
+  _audience = None
+
+  def __init__(self, token=None, expiry=None, *args, **kwargs):
+    super(*args, **kwargs)
+    self.expiry = expiry
+    self.token = token
+
+  @property
+  def info(self):
+    return {}
+
+
+class TestWrappedCredentials(testcase.GsUtilUnitTestCase):
+  """Test logic for interacting with Wrapped Credentials."""
+
+  @mock.patch("oauth2client.transport.httplib2.Http", autospec=True)
+  def testWrappedCredentialUsage(self, http):
+    http.return_value.request.return_value = (RESPONSE, CONTENT)
+
+    creds = WrappedCredentials(MockCredentials(token=ACCESS_TOKEN))
+
+    http = oauth2client.transport.get_http_object()
+    creds.authorize(http)
```

Comment:
It's not really clear what we are trying to test here. The creds instance is not getting used as you might expect.
i.e I think even if we remove this line, this test will still pass. So we are not really testing any of the WrappedCredentials functionality.

It might be worth testing the specific methods in this class, because looks like that's where most of the logic is present. e.g to_json, from_json, for_external_account etc. Let me know if you need any help on this.

Reply:
I updated the test to make sure that requests being sent actually use the information we set in the mocked credentials.  I also added some tests for to_json and from_json.  Also, while testing, I noticed that WrappedCredentials doesn't *really* work with any type of credentials; at least not if we intend serialization to work, because from_json assumes that we are using either AWS or IdentityPool based credentials, so I've updated some of WrappedCredentials accordingly